infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,605 @@
+                    requestID);
+        }
+    }
+
+    @Override
+    public void stop(Set<LogReplicationSession> sessions) {
+        sessions.forEach(session -> removeSessionInfo(session));
+
+        // When all the connection starter sessions are stopped, shut down the transport layer
+        if(sessionToOutstandingRequests.isEmpty() && this.clientChannelAdapter != null) {
+            // stop the client Adapter. The Server Adapter is closed in the interClusterServerNode.
+            this.clientChannelAdapter.stop();
+        }
+    }
+
+    @Override
+    public void setTimeoutResponse(long timeoutResponse) {
+        this.timeoutResponse = timeoutResponse;
+    }
+
+    /**
+     * Interfaces between the LogReplicationServer (msg handler), and the transport layer
+     * @param response Log replication response message
+     */
+    @Override
+    public void sendResponse(CorfuMessage.ResponseMsg response) {
+        log.info("Ready to send response {}", response.getPayload().getPayloadCase());
+        LogReplicationSession session = response.getHeader().getSession();
+        if (isConnectionInitiator(session)) {
+            if(sessionManager.getIncomingSessions().contains(session)) {
+                if (sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().isPresent()) {
+                    clientChannelAdapter.send(sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().get(), response);
+                } else {
+                    sessionToSinkVerifyLeadership.get(session).input(
+                            new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS));
+                }
+            }
+        } else {
+            try {
+                this.serverChannelAdapter.send(response);
+                log.info("Sent response: {}", response);
+            } catch (IllegalArgumentException e) {
+                log.warn("Illegal response type. Ignoring message.", e);
+            }
+        }
+    }
+
+    /**
+     * Receive Corfu Message from the Channel Adapter for further processing
+     *
+     * @param msg received corfu message
+     */
+    @Override
+    public void receive(CorfuMessage.ResponseMsg msg) {
+        try {
+            LogReplicationSession session = msg.getHeader().getSession();
+            String nodeId = msg.getPayload().getLrLeadershipLoss().getNodeId();
+
+            // If it is a Leadership Loss Message re-trigger leadership discovery
+            if (msg.getPayload().getPayloadCase() == CorfuMessage.ResponsePayloadMsg.PayloadCase.LR_LEADERSHIP_LOSS) {
+                if (isOutgoingSession(session)) {
+                    this.sessionToRuntimeFSM.get(session)
+                            .input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS, nodeId));
+                } else {
+                    sessionToSinkVerifyLeadership.get(session)
```

Comment:
looks like this is simulating a state machine just like on the Source.  Is it a good idea to reuse the Source's state machine then?

Reply:
true, it does look similar. But if we have it in the source's state machine, we would end up having a lot of "if"s since the behaviour is slightly different.
The logic to trigger leadership is the same, but mostly thats the only part.